src/core/main/common/FreezeResolver.ts

```diff
@@ -0,0 +1,573 @@
+   * an object describing this wanted unfreezing strategy.
+   *
+   * If this method decides to take no action for now, it returns `null`.
+   * @param {Object} observation - playback observation that has just been
+   * performed.
+   * @param {number} now - Monotonically-raising timestamp for the current
+   * time.
+   * @returns {Object|null}
+   */
+  private _checkForDecipherabilityRelatedFreeze(
+    observation: IFreezeResolverObservation,
+    now: number,
+  ): IFreezeResolution | null {
+    const { readyState, rebuffering, freezing, fullyLoaded } = observation;
+    const bufferGap = normalizeBufferGap(observation.bufferGap);
+    const rebufferingForTooLong =
+      rebuffering !== null && now - rebuffering.timestamp > 4000;
+    const frozenForTooLong = freezing !== null && now - freezing.timestamp > 4000;
+
+    const hasDecipherabilityFreezePotential =
+      (rebufferingForTooLong || frozenForTooLong) &&
+      (bufferGap >= 6 || fullyLoaded) &&
+      readyState <= 1;
+
+    if (!hasDecipherabilityFreezePotential) {
+      this._decipherabilityFreezeStartingTimestamp = null;
+    } else if (this._decipherabilityFreezeStartingTimestamp === null) {
+      log.debug("FR: Start of a potential decipherability freeze detected");
+      this._decipherabilityFreezeStartingTimestamp = now;
+    }
+
+    const shouldHandleDecipherabilityFreeze =
+      this._decipherabilityFreezeStartingTimestamp !== null &&
+      getMonotonicTimeStamp() - this._decipherabilityFreezeStartingTimestamp > 4000;
+
+    let hasOnlyDecipherableSegments = true;
+    let isClear = true;
+    for (const ttype of ["audio", "video"] as const) {
+      const status = this._segmentSinksStore.getStatus(ttype);
+      if (status.type === "initialized") {
+        for (const segment of status.value.getLastKnownInventory()) {
+          const { representation } = segment.infos;
+          if (representation.decipherable === false) {
+            log.warn("FR: we have undecipherable segments left in the buffer, reloading");
+            this._decipherabilityFreezeStartingTimestamp = null;
+            this._ignoreFreezeUntil = now + MINIMUM_TIME_BETWEEN_FREEZE_HANDLING;
+            return { type: "reload", value: null };
+          } else if (representation.contentProtections !== undefined) {
+            isClear = false;
+            if (representation.decipherable !== true) {
+              hasOnlyDecipherableSegments = false;
+            }
+          }
+        }
+      }
+    }
+
+    if (shouldHandleDecipherabilityFreeze && !isClear && hasOnlyDecipherableSegments) {
+      log.warn(
+        "FR: we are frozen despite only having decipherable " +
+          "segments left in the buffer, reloading",
+      );
+      this._decipherabilityFreezeStartingTimestamp = null;
+      this._ignoreFreezeUntil = now + MINIMUM_TIME_BETWEEN_FREEZE_HANDLING;
+      return { type: "reload", value: null };
+    }
+    return null;
+  }
```

Comment:
one thing I'm not understanding is that a reload is done only if there is decipherable segments in the buffer?
I though we were reloading because there was a segment left in the buffer that is NOT decipherable

Reply:
We're (and we were already) doing both. Higher up in the code we also return `"reload"` if a `decipherable === false` is encountered.

The "reload even though all segments are known to be decipherable" thing is done to work around a known recurrent PlayReady issue.
And on that part, I don't think we can just expect them to fix it once and for all, so it's kind of a "well, EME is broken again" resilience code.